test/type/nested_types.swift

```diff
@@ -84,3 +84,16 @@ do {
   let _: (Int, Int).Undef // expected-error {{'Undef' is not a member type of type '(Swift.Int, Swift.Int)'}}
   let _: ((Int) -> Void).Undef // expected-error {{'Undef' is not a member type of type '(Swift.Int) -> Swift.Void'}}
 }
+
+// rdar://106446740 - source compatibility break - reference to inner type without `.self` used to work
+func test_inner_types_dont_require_self() {
+  struct Test {
+    enum E {}
+
+    func decode<T>(_: T.Type) {}
+
+    func test() {
+      decode(Self.E) // Ok
```

Comment:
I think this is an acceptable source break since this wasn't supposed to work previously. We can document this in the CHANGELOG.

Reply:
If the only case that was broken before was when the base type is written as `Self`, maybe MiscDiagnostics can downgrade to a warning if the TypeRepr is written that way, and then we can turn it into an error in -swift-version 6.

Reply:
```swift
struct Outer {
  struct Inner {}

  func f() {
    print(Outer.Inner)
    print(Self.Inner)
  }
}
```

Yeah, we used to reject the first print() but accept the second one.

Reply:
godbolt is telling me this behavior has been around since 5.2. Probably when we started mapping `Self` in expression context to a `FixedTypeRepr` rather than an `IdentTypeRepr`.

Reply:
It’s worse than just `Self`. All of these used to work:

```swift
extension Optional {
  typealias Bummer = Never
}
do {
  func decode<T>(_: T.Type) {}

  decode([Int].Element)
  decode([Int : Int].Element)
  decode(Int?.Bummer)
  decode((Int?).Bummer)
}
```

Reply:
Can we change MiscDiagnostics to look at the TypeRepr, and only emit the error about missing .self if the TypeRepr has the original form where each component is a non-Self identifier? For anything else, only emit the error with -swift-version 6.

Reply:
Yeah that sounds good. Basically warn if the base component is not a `IdentTypeRepr`.

Reply:
Hoping to have a fix before the bots finish.

Reply:
#64229